In [1]:
import pandas as pd

data = pd.read_csv('static/real-case.csv', header=None)
data

/home/visus/.pyenv/versions/3.7.0/lib/python3.7/site-packages/pandas/compat/__init__.py:117: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs               NaN   
2           chutney                NaN          NaN               NaN   
3            turkey            avocado          NaN               NaN   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread               NaN   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken                NaN          NaN               NaN   
7499       escalope          green tea          NaN               NaN   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1              NaN               NaN   NaN             NaN           NaN   
2              NaN               NaN   NaN             NaN           NaN   
3              NaN               NaN   NaN             NaN           NaN   
4        green tea               NaN   NaN             NaN           NaN   
...            ...               ...   ...             ...           ...   
7496           NaN               NaN   NaN             NaN           NaN   
7497     magazines         green tea   NaN             NaN           NaN   
7498           NaN               NaN   NaN             NaN           NaN   
7499           NaN               NaN   NaN             NaN           NaN   
7500           NaN               NaN   NaN             NaN           NaN   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1              NaN             NaN        NaN    NaN    NaN            NaN   
2              NaN             NaN        NaN    NaN    NaN            NaN   
3              NaN             NaN        NaN    NaN    NaN            NaN   
4              NaN             NaN        NaN    NaN    NaN            NaN   
...            ...             ...        ...    ...    ...            ...   
7496           NaN             NaN        NaN    NaN    NaN            NaN   
7497           NaN             NaN        NaN    NaN    NaN            NaN   
7498           NaN             NaN        NaN    NaN    NaN            NaN   
7499           NaN             NaN        NaN    NaN    NaN            NaN   
7500           NaN             NaN        NaN    NaN    NaN            NaN   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1        NaN                NaN              NaN      NaN        NaN  
2        NaN                NaN              NaN      NaN        NaN  
3        NaN                NaN              NaN      NaN        NaN  
4        NaN                NaN              NaN      NaN        NaN  
...      ...                ...              ...      ...        ...  
7496     NaN                NaN              NaN      NaN        NaN  
7497     NaN                NaN              NaN      NaN        NaN  
7498     NaN                NaN              NaN      NaN        NaN  
7499     NaN                NaN              NaN      NaN        NaN  
7500     NaN                NaN              NaN      NaN        NaN  

[7501 rows x 20 columns]

* ## The above table is showing the transactions for a week at a market in French



In [19]:
data = pd.read_csv('static/real-case.csv', header = None)
transactions = []
for i in range(0, len(data)):
    transactions.append([str(data.values[i, j]) for j in range(0, len(data.columns))])
transactions[0:3]

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers',
  'meatballs',
  'eggs',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['chutney',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan']]

* __An observation before continue coding:__
>    To return a considerable number of elements, I needed to set a much smaller number to my_support and to my_confidence

In [48]:
from apyori import apriori
rules = apriori(transactions, 
                     min_support = 0.003,
                     min_confidence = 0.2,
                     min_lift = 3.0,
                     min_lenght = 2)
rules = list(rules)
len(rules)

160

In [49]:
results = list(rules)
results[0:5]

[RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [50]:
results2 = [list(x) for x in results]
results2[0:5]

[[frozenset({'chicken', 'light cream'}),
  0.004532728969470737,
  [OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]],
 [frozenset({'escalope', 'mushroom cream sauce'}),
  0.005732568990801226,
  [OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]],
 [frozenset({'escalope', 'pasta'}),
  0.005865884548726837,
  [OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]],
 [frozenset({'fromage blanc', 'honey'}),
  0.003332888948140248,
  [OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0.2450980392156863, lift=5.164270764485569)]],
 [frozenset({'ground beef', 'herb & pepper'}),
  0.015997866951073192,
  [OrderedStatistic(items_base=frozenset({'herb & peppe

In [51]:
formatedresults = []
for j in range(0, len(results2)):
    formatedresults.append([list(x) for x in results2[j][2]])
formatedresults[0:5]

[[[frozenset({'light cream'}),
   frozenset({'chicken'}),
   0.29059829059829057,
   4.84395061728395]],
 [[frozenset({'mushroom cream sauce'}),
   frozenset({'escalope'}),
   0.3006993006993007,
   3.790832696715049]],
 [[frozenset({'pasta'}),
   frozenset({'escalope'}),
   0.3728813559322034,
   4.700811850163794]],
 [[frozenset({'fromage blanc'}),
   frozenset({'honey'}),
   0.2450980392156863,
   5.164270764485569]],
 [[frozenset({'herb & pepper'}),
   frozenset({'ground beef'}),
   0.3234501347708895,
   3.2919938411349285]]]

## How to define the min values to these params?

* __Example:__ 

    `I am looking for an item which is sold 3 times a day. In this case, how many times is this item sold in week?`

* __Therefore, for support:__

>3(times sold) * 7(days at week) = 21
>21 / 7501(number of transactions in the list) = 2,79 * 10^-3 or 0,00279

* __This idea suggests that the greater the number of records, the lesser the support and confidence__

# An idea that could be applied:
* __From Apyori results:__
    >`frozenset({'light cream'}),`
    >`frozenset({'chicken'}),`
    >`0.29059829059829057,`
    >`4.84395061728395`

* As we can notice, the people whose buy light cream has almost 5 times more tendence to buy chicken in the same order than the people don't buy light cream. So, it could be a good idea to place these two items near themselves.